remove all raw_inputs and change params handling

PeterMitrano opened the pull request:
- now using rospkg to find default_params, instead of `__file__`
 - no prompting with raw_input
 - adds dependancy on pathlib

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -1,12 +1,13 @@
 from os.path import join
 
 import git
+import rospkg
```
I had originally wrote shape completion training such that there was no dependency on ROS.
I'm okay throwing that out the window now. It has a cost, but oh well

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,79 +18,65 @@ def unique_trial_name(*names):
-    default_keys = set(default_params.keys())
-
-    if given_keys != default_keys:
-        print()
```
Add
```
if not given_keys <= default_keys:
  print("!! Warning !!"))
  print("hyper parameters were specified that do not have default values. All hyperparameters should have default values to allow for proper reloading of older models")
  for missing_default_key in given_keys - default_keys:
    print("{} missing from defaults".format(missing_default_key))
```


This was a reminder to keep my `default_params.json` up to date.
I'm okay updating to the system we discussed. i.e.
- you do not need to (and should not) pass in every param for every model.
- There is a `default_params.json` that has a default value for every hparam we might use

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,79 +18,65 @@ def unique_trial_name(*names):
-            print("{} missing from specified keys".format(missing_given_key))
-        print("Press any key to continue")
-        raw_input()
+    model_params = json.load(open(join(trial_directory, "params.json"), 'r'))
```
"params.json" might not exist (and wont exist unless reloading)

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,79 +18,65 @@ def unique_trial_name(*names):
+    model_params = json.load(open(join(trial_directory, "params.json"), 'r'))
+    return _handle_params(model_params, given_params)
 
-    _write_params(model_params_fp, given_params)
```
Where do you handle writing the params file now?
Some params file needs to be written for new trials

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,79 +18,65 @@ def unique_trial_name(*names):
-    return given_params
 
+def _handle_params(model_params, given_params):
+    model_params.update(given_params)
```
I think we need to discuss the role of "default", "model", and "given" params. I don't think we need "given" any more, if we are copying a json file.
"default" : straightforward
"model": params different from defaults used to train the model


```
def _handle_params(model_hparams, override_hparams):
  params = get_default_params()
  params.update(model_params)
  return params
```

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,79 +18,65 @@ def unique_trial_name(*names):
 
 
-def handle_params(default_params_fp, model_params_fp, given_params):
+def handle_params(trial_directory, given_params):
```
This needs some more work.
the docstring no longer holds, and this will not work with the existing architecture.

Somehow this (or a similar function) needs to 
```
if creating new model_model:
    choose which params to use (json, python dict, ???)
    write a json file to a model directory
return load(existing params)

```

PeterMitrano commented:
note for when I work on this next, I also need to implement the implicit save / restore behavior
also add more tests to show that the things brad mentioned are current wrong/broken get fixed

PeterMitrano committed:
more work on filepath refactor

PeterMitrano committed:
finish merging

PeterMitrano committed:
move default params

PeterMitrano commented:
this thing keeps getting bigger sorry :upside_down_face:

bsaund commented:
No problem! It's not that long yet

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,108 +18,99 @@ def unique_trial_name(*names):
 
-def get_trial_directory(base_directory, nick=None, expect_reuse=False, write_summary=True):
+def create_or_load_trial(trial_name=None, params=None, trials_directory=None, write_summary=True):
+    if '/' in trial_name:
```
1) I'm confused that `training` seems to be equivalent to "loading existing network". What if we want to continue training an existing network?


2) Cleanup logic (though this will change based on the previous comment)
```
training = '/' in trial_name

if training:
  if params is not None:
    print("Warning! params given but not used, since loading existing trial)
  full_trial_directory, params = load_trial(trial_name=trial_name, trials_directory=trials_directory)
else:
  full_trial_directory = create_trial(params=params,
                                            trial_name=trial_name,
                                            write_summary=write_summary,
                                            trials_directory=trials_directory)
all_params = get_default_params()
all_params.update(params)
```

PeterMitrano replied:
hmm yea I think we cannot infer if the user wants to train or not -- only whether they want to load or not. if there's a '/' in the name we might just be loading because they want to do evaluation. So I say we remove the training return value entirely.

from what I can tell the training return is only used in model runner to set the batch size to 1 vs 16

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,108 +18,99 @@ def unique_trial_name(*names):
+    if '/' in trial_name:
+        training = False
+        # load, raises an exception if it doesn't exist
+        full_trial_directory, params = load_trial(trial_name=trial_name, trials_directory=trials_directory)
```
warn if `params is not None` before this

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,108 +18,99 @@ def unique_trial_name(*names):
-
-    given_keys = set(given_params.keys())
-    default_keys = set(default_params.keys())
+        _write_summary(full_directory, trial_name)
```
This is odd (wrong) behavior when `trial_name is None`

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,108 +18,99 @@ def unique_trial_name(*names):
+        trials_directory = shape_completion_training_path / 'trials'
+    trials_directory.mkdir(parents=True, exist_ok=True)
+
+    if trial_name is None:
```
Use only one variable. `trial_name` or `group_name`.

bsaund commented on shape_completion_training/src/shape_completion_training/model/filepath_tools.py:
```diff
@@ -17,108 +18,99 @@ def unique_trial_name(*names):
+        print("No trial_namename given. Using tmp")
+        group_name = "tmp"
+    else:
+        group_name = trial_name
```
This logic is cleaner:

```
if trial_name is None:
  print("No trial_name given. Using tmp")
  trial_name = "tmp"

(no else statement needed)
```

bsaund commented on shape_completion_training/src/shape_completion_training/tests/test_filepath_tools.py:
```diff
@@ -0,0 +1,56 @@
+    def test_get_trial_directory_new_train_error(self):
+        base_dir = pathlib.Path('.testing')
+        unique_nickname = 'new_trial/subdir_non_existing'
+        try:
```
```
self.assertRaises(ValueError, filepath_tools.create_or_load_trial(trial_name=unique_nickname,
                                                params={},
                                                write_summary=False,
                                                trials_directory=base_dir)
```

bsaund reviewed (approved):
Go ahead and fix what you want and merge after that. This will break some of my stuff (e.g. at least the trial directory structure), so I might as well start fixing it

PeterMitrano committed:
address comments

still not sure how to handle ambiguous case where the user provides only
the group name but intends to load an existing model. In this case I
think what currently will happen is that a new directory/model will be
made? we need to have code that checks to make sure a checkpoint exists
before doing evaluation or something

PeterMitrano commented:
I'm gonna wait to merge this until we agree on what we want to happen when the user provides only a group name but intended to load an existing model.

PeterMitrano commented:
TODO new behavior:

two strings passed to `create_or_load_trial`, one `group_name`, one `trial_path`
if you pass only group name is makes a new trial
if you pass only trial path it loads from that
if you pass none it creates a new tmp trial
if you pass path it warns and ignores group_name

PeterMitrano committed:
update to match new expected behavior with "group_name" and "trial_path"

PeterMitrano committed:
corresponding updates to model runner and usage of model runner

PeterMitrano commented:
this change breaks how model runner is used in train.py

bsaund commented:
I'm merging now. 
I'm using the neural nets in inference today, so I'll see what breaks :)
